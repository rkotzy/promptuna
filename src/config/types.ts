import type { PromptunaObservability } from '../observability/types';
import type { Message } from '../templates/types';
import type { ResponseFormat, FallbackTarget, ModelParams } from '../responses/types';
import type { Routing } from '../routing/types';
import type { ChatMessage } from '../providers/types';

export interface PromptunaConfig {
  version: string;
  providers: Record<string, ProviderConfig>;
  responseSchemas?: Record<string, any>;
  prompts: Record<string, Prompt>;
}

export interface PromptunaRuntimeConfig {
  configPath: string;
  openaiApiKey?: string;
  anthropicApiKey?: string;
  googleApiKey?: string;
  environment?: 'dev' | 'prod';
  onObservability?: (event: PromptunaObservability) => void;
}

export interface ProviderConfig {
  type: 'openai' | 'anthropic' | 'google';
  config?: Record<string, any>; // provider-specific additional config
}

export interface Variant {
  provider: string;
  model: string;
  default?: boolean;
  parameters?: ModelParams;
  messages: Message[];
  responseFormat?: ResponseFormat;
  fallback?: FallbackTarget[];
}

export interface Prompt {
  description: string;
  variants: Record<string, Variant>;
  routing: Routing;
  chains?: any[]; // loosely typed for now
}

export interface ValidationResult {
  valid: boolean;
  errors?: Array<{
    message: string;
    path: string;
    keyword: string;
  }>;
}

// Error types
export class PromptunaError extends Error {
  constructor(
    message: string,
    public code: string,
    public details?: any
  ) {
    super(message);
    this.name = 'PromptunaError';
  }
}

export class ConfigurationError extends PromptunaError {
  constructor(message: string, details?: any) {
    super(message, 'CONFIGURATION_ERROR', details);
  }
}

export class ExecutionError extends Error {
  constructor(
    message: string,
    public details?: any
  ) {
    super(message);
    this.name = 'ExecutionError';
  }
}

// SDK API parameter types
export interface ChatCompletionParams {
  /** The ID of the prompt */
  promptId: string;
  /** Variables to interpolate into the template */
  variables?: Record<string, any>;
  /** Conversation history to prepend to the prompt (e.g., previous messages) */
  messageHistory?: ChatMessage[];
  /** Stable identifier used to hash into deterministic traffic buckets */
  userId?: string;
  /** Tags that describe the request context (e.g., geography, experiment flags) */
  tags?: string[];
  /** Unix timestamp (seconds) to evaluate phased roll-outs. Defaults to now. */
  unixTime?: number;
}

export interface GetTemplateParams {
  /** The ID of the prompt */
  promptId: string;
  /** The ID of the variant within the prompt */
  variantId: string;
  /** Variables to interpolate into the template */
  variables?: Record<string, any>;
}