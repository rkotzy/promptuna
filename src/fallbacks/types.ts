import type { ProviderError } from '../errors';

export interface ExecutionTarget {
  providerId: string; // Key into config.providers
  providerType: string; // e.g. "openai", "anthropic", "google"
  model: string;
}

export interface FallbackCallbackContext {
  target: ExecutionTarget;
  error?: ProviderError;
}