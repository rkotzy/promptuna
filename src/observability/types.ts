export interface PromptunaObservability {
  // Request identification
  requestId: string;
  /** Stable identifier for the end-user making the request */
  userId?: string;
  /** ISO-8601 timestamp in UTC */
  timestamp: string;
  /** Version of the Promptuna SDK (e.g. read from package.json) */
  sdkVersion: string;
  /** Runtime environment – useful when the same service is deployed to multiple envs */
  environment?: 'dev' | 'prod';

  // Prompt and variant selection
  promptId: string;
  /** The concrete variant executed – renamed from selectedVariant */
  variantId: string;
  routingReason: RoutingReason;
  routingTags?: string[];

  // Performance metrics
  timings: {
    /** End-to-end duration (ms) */
    total: number;
    /** Time spent rendering templates (ms) */
    template?: number;
    /** Time spent waiting for the LLM provider (ms) */
    provider?: number;
    /** Number of automatic retries that occurred */
    retries?: number;
  };
  tokenUsage?: TokenUsage;

  // Provider information
  provider: string;
  model: string;
  /** Provider-side request / response identifier for easier cross-referencing */
  providerRequestId?: string;

  // Fallback tracking
  fallbackUsed: boolean;
  /** Full chain of fallback attempts – first element is the initial attempt when fallbackUsed is true */
  fallbacks?: Array<{
    provider: string;
    model: string;
    reason: 'provider-error' | 'timeout' | 'rate-limit';
  }>;

  // Request outcome
  success: boolean;
  error?: ObservabilityError;

  // A/B testing context
  experimentContext?: {
    tags: string[];
    weightedSelection: boolean;
    selectedWeight: number;
  };

  // Phased rollout context
  phasedRolloutContext?: {
    phase: string;
    rolloutPercentage: number;
    rolloutStart: number;
    rolloutEnd?: number;
  };

  /** Free-form bucket for application-specific metadata */
  custom?: Record<string, any>;
}

export interface TokenUsage {
  prompt: number;
  completion: number;
  total: number;
}

export interface ObservabilityError {
  type: string;
  message: string;
  code?: string;
  retryable: boolean;
  provider?: string;
  httpStatus?: number;
  stack?: string;
}

export type RoutingReason =
  | 'tag-match'
  | 'phased-rollout'
  | 'weight-distribution'
  | 'default';

export interface Timings {
  /** End-to-end duration – automatically filled */
  total: number;
  /** Time spent rendering Liquid templates */
  template?: number;
  /** Time spent waiting for provider response */
  provider?: number;
  /** Number of retries that occurred */
  retries?: number;
  // eslint-disable-next-line @typescript-eslint/ban-types
  [key: string]: number | undefined;
}

export interface BuilderInit {
  sdkVersion: string;
  environment?: 'dev' | 'prod';
  promptId: string;
  userId?: string;
  /** Variant may be unknown at construction time */
  variantId?: string;
  routingReason: PromptunaObservability['routingReason'];
  /** Tags that influenced routing (if any) */
  routingTags?: string[];
  /** Experiment context for weighted selection */
  experimentContext?: PromptunaObservability['experimentContext'];
  emit?: (evt: PromptunaObservability) => void;
}
