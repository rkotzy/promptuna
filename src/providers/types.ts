export interface ChatMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

import type { ResponseFormat } from '../responses/types';

export interface ChatCompletionOptions {
  messages: ChatMessage[];
  model: string;
  userId?: string;
  responseFormat?: ResponseFormat;
  responseSchema?: any; // Resolved JSON schema for json_schema type
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  frequency_penalty?: number;
  presence_penalty?: number;
  stop?: string | string[];
  logit_bias?: Record<string, number>;
  [key: string]: any; // Allow arbitrary parameters since we will transform these into provider-specific params
}

export interface ChatCompletionResponse {
  id: string;
  model: string;
  choices: Array<{
    message: ChatMessage;
    finish_reason: string | null;
    index: number;
  }>;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

export interface Provider {
  chatCompletion(
    options: ChatCompletionOptions
  ): Promise<ChatCompletionResponse>;
}
