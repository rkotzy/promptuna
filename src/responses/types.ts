export interface ResponseFormat {
  type: 'json_schema' | 'raw_text';
  schemaRef?: string;
}

export interface FallbackTarget {
  provider: string;
  model: string;
}

export interface ModelParams {
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  frequency_penalty?: number;
  presence_penalty?: number;
  stop?: string | string[];
  json_mode?: boolean;
  top_k?: number;
  candidate_count?: number;
  logit_bias?: Record<string, number>;
}