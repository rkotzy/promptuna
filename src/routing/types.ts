import type { Variant, Prompt } from '../config/types';
import type { RoutingReason } from '../observability/types';

export interface VariantSelection {
  variantId: string;
  variant: Variant;
  reason: RoutingReason;
  weightPicked?: number;
}

export interface VariantSelectorParams {
  prompt: Prompt;
  promptId: string;
  userId?: string;
  tags?: string[];
  /** Unix timestamp in seconds */
  now?: number;
}

export interface RoutingRule {
  target: string;
  weight?: number;
  tags?: string[];
}

export interface PhasedRule {
  start: number;
  end?: number;
  weights: Record<string, number>;
}

export interface Routing {
  rules: RoutingRule[];
  phased?: PhasedRule[];
}
