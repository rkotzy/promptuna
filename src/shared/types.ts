export type ProviderId = 'openai' | 'anthropic' | 'google';

/**
 * Mapping rules for a single parameter for a specific provider.
 * - `param`   : name that the provider SDK expects
 * - `min`/`max`: optional clamp range (after any scale)
 * - `scale`   : optional transform applied to the canonical value before clamping
 * If the value is `false`, the parameter is not supported for that provider.
 */
export type MappingRule =
  | {
      param: string;
      min?: number;
      max?: number;
      scale?: (v: any) => any;
    }
  | false;
