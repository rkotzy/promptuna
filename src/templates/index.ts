export { TemplateProcessor } from './processor';
export type { Message, RenderedMessage, TemplateError } from './types';
