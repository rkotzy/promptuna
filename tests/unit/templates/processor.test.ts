import { describe, it, expect, vi, beforeEach } from 'vitest';
import { TemplateProcessor } from '../../../src/templates/processor';
import { TemplateError } from '../../../src/templates/types';

describe('TemplateProcessor', () => {
  let processor: TemplateProcessor;

  beforeEach(() => {
    processor = new TemplateProcessor();
  });

  describe('processTemplate', () => {
    it('should process a simple template with variables', async () => {
      const template = 'Hello {{name}}!';
      const variables = { name: 'Alice' };

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('Hello Alice!');
    });

    it('should process templates with multiple variables', async () => {
      const template = 'Hello {{name}}, welcome to {{city}}!';
      const variables = { name: 'Alice', city: 'San Francisco' };

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('Hello Alice, welcome to San Francisco!');
    });

    it('should handle missing variables gracefully', async () => {
      const template = 'Hello {{name}}, your score is {{score}}!';
      const variables = { name: 'Alice' };

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('Hello Alice, your score is !');
    });

    it('should handle nested object variables', async () => {
      const template = 'Hello {{user.name}}, your email is {{user.email}}!';
      const variables = { 
        user: { 
          name: 'Alice', 
          email: 'alice@example.com' 
        } 
      };

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('Hello Alice, your email is alice@example.com!');
    });

    it('should handle arrays in templates', async () => {
      const template = 'Items: {{items}}';
      const variables = { items: ['apple', 'banana', 'cherry'] };

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('Items: applebananacherry');
    });

    it('should handle conditional logic', async () => {
      const template = '{% if user %}Hello {{user}}!{% else %}Hello guest!{% endif %}';
      
      const resultWithUser = await processor.processTemplate(template, { user: 'Alice' });
      expect(resultWithUser).toBe('Hello Alice!');

      const resultWithoutUser = await processor.processTemplate(template, {});
      expect(resultWithoutUser).toBe('Hello guest!');
    });

    it('should handle loops', async () => {
      const template = '{% for item in items %}{{item}} {% endfor %}';
      const variables = { items: ['apple', 'banana', 'cherry'] };

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('apple banana cherry ');
    });

    it('should throw TemplateError for invalid template syntax', async () => {
      const template = 'Hello {{name}'; // Missing closing brace
      const variables = { name: 'Alice' };

      await expect(processor.processTemplate(template, variables))
        .rejects.toThrow(TemplateError);
    });

    it('should include helpful error context in TemplateError', async () => {
      const template = 'Hello {{name}'; // Missing closing brace
      const variables = { name: 'Alice' };

      try {
        await processor.processTemplate(template, variables);
        expect.fail('Should have thrown a TemplateError');
      } catch (error) {
        expect(error).toBeInstanceOf(TemplateError);
        expect((error as TemplateError).details).toMatchObject({
          template: 'Hello {{name}',
          variables: ['name'],
        });
      }
    });

    it('should handle empty templates', async () => {
      const template = '';
      const variables = {};

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('');
    });

    it('should handle templates with only text', async () => {
      const template = 'This is just plain text';
      const variables = {};

      const result = await processor.processTemplate(template, variables);

      expect(result).toBe('This is just plain text');
    });
  });

  describe('custom filters', () => {
    describe('join filter', () => {
      it('should join arrays with default separator', async () => {
        const template = '{{items | join}}';
        const variables = { items: ['apple', 'banana', 'cherry'] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('apple, banana, cherry');
      });

      it('should join arrays with custom separator', async () => {
        const template = '{{items | join: " | "}}';
        const variables = { items: ['apple', 'banana', 'cherry'] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('apple | banana | cherry');
      });

      it('should handle non-array values', async () => {
        const template = '{{value | join}}';
        const variables = { value: 'not an array' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('not an array');
      });

      it('should handle empty arrays', async () => {
        const template = '{{items | join}}';
        const variables = { items: [] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('');
      });
    });

    describe('numbered filter', () => {
      it('should number array items with default prefix', async () => {
        const template = '{{items | numbered}}';
        const variables = { items: ['apple', 'banana', 'cherry'] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('  1. apple  2. banana  3. cherry');
      });

      it('should number array items with custom prefix', async () => {
        const template = '{{items | numbered: "- "}}';
        const variables = { items: ['apple', 'banana', 'cherry'] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('- 1. apple- 2. banana- 3. cherry');
      });

      it('should handle non-array values', async () => {
        const template = '{{value | numbered}}';
        const variables = { value: 'not an array' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('not an array');
      });

      it('should handle empty arrays', async () => {
        const template = '{{items | numbered}}';
        const variables = { items: [] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('');
      });
    });

    describe('default filter', () => {
      it('should use default value for null', async () => {
        const template = '{{value | default: "fallback"}}';
        const variables = { value: null };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('fallback');
      });

      it('should use default value for undefined', async () => {
        const template = '{{value | default: "fallback"}}';
        const variables = {};

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('fallback');
      });

      it('should use default value for empty string', async () => {
        const template = '{{value | default: "fallback"}}';
        const variables = { value: '' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('fallback');
      });

      it('should use original value when not null/undefined/empty', async () => {
        const template = '{{value | default: "fallback"}}';
        const variables = { value: 'original' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('original');
      });

      it('should handle zero as a valid value', async () => {
        const template = '{{value | default: "fallback"}}';
        const variables = { value: 0 };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('0');
      });

      it('should handle false as a valid value', async () => {
        const template = '{{value | default: "fallback"}}';
        const variables = { value: false };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('false');
      });
    });

    describe('string case filters', () => {
      it('should capitalize first letter', async () => {
        const template = '{{value | capitalize}}';
        const variables = { value: 'hello world' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('Hello world');
      });

      it('should convert to lowercase', async () => {
        const template = '{{value | downcase}}';
        const variables = { value: 'HELLO WORLD' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('hello world');
      });

      it('should convert to uppercase', async () => {
        const template = '{{value | upcase}}';
        const variables = { value: 'hello world' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('HELLO WORLD');
      });

      it('should handle non-string values gracefully', async () => {
        const template = '{{value | capitalize}}';
        const variables = { value: 123 };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('123');
      });

      it('should handle empty strings', async () => {
        const template = '{{value | capitalize}}';
        const variables = { value: '' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('');
      });
    });

    describe('size filter', () => {
      it('should return array length', async () => {
        const template = '{{items | size}}';
        const variables = { items: ['apple', 'banana', 'cherry'] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('3');
      });

      it('should return string length', async () => {
        const template = '{{text | size}}';
        const variables = { text: 'hello world' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('11');
      });

      it('should return object key count', async () => {
        const template = '{{obj | size}}';
        const variables = { obj: { a: 1, b: 2, c: 3 } };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('3');
      });

      it('should return 0 for null/undefined', async () => {
        const template = '{{value | size}}';
        const variables = { value: null };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('0');
      });

      it('should return 0 for empty arrays', async () => {
        const template = '{{items | size}}';
        const variables = { items: [] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('0');
      });

      it('should return 0 for empty objects', async () => {
        const template = '{{obj | size}}';
        const variables = { obj: {} };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('0');
      });
    });

    describe('chained filters', () => {
      it('should handle multiple filters in sequence', async () => {
        const template = '{{items | join: " " | upcase}}';
        const variables = { items: ['hello', 'world'] };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('HELLO WORLD');
      });

      it('should handle complex filter chains', async () => {
        const template = '{{value | default: "empty" | capitalize | size}}';
        const variables = { value: '' };

        const result = await processor.processTemplate(template, variables);

        expect(result).toBe('5'); // "Empty" has 5 characters
      });
    });
  });

  describe('complex template scenarios', () => {
    it('should handle mixed template with variables, conditions, and filters', async () => {
      const template = `
{% if user %}
Hello {{user.name | capitalize}}!
{% if user.items %}
Your items: {{user.items | join: ", "}}
Total: {{user.items | size}} items
{% endif %}
{% else %}
Welcome, guest!
{% endif %}
      `.trim();

      const variables = {
        user: {
          name: 'alice',
          items: ['apple', 'banana', 'cherry']
        }
      };

      const result = await processor.processTemplate(template, variables);

      expect(result).toContain('Hello Alice!');
      expect(result).toContain('Your items: apple, banana, cherry');
      expect(result).toContain('Total: 3 items');
    });

    it('should handle nested loops and conditions', async () => {
      const template = `
{% for category in categories %}
{{category.name | upcase}}:
{% for item in category.items %}
  - {{item | capitalize}}
{% endfor %}
{% endfor %}
      `.trim();

      const variables = {
        categories: [
          { name: 'fruits', items: ['apple', 'banana'] },
          { name: 'vegetables', items: ['carrot', 'broccoli'] }
        ]
      };

      const result = await processor.processTemplate(template, variables);

      expect(result).toContain('FRUITS:');
      expect(result).toContain('  - Apple');
      expect(result).toContain('  - Banana');
      expect(result).toContain('VEGETABLES:');
      expect(result).toContain('  - Carrot');
      expect(result).toContain('  - Broccoli');
    });

    it('should handle complex message formatting', async () => {
      const template = `
{{greeting | default: "Hello"}}, {{name | capitalize}}!

{% if tasks %}
Here are your tasks ({{tasks | size}} total):
{{tasks | numbered: "  "}}
{% else %}
No tasks for today.
{% endif %}

Best regards,
{{sender | default: "System"}}
      `.trim();

      const variables = {
        name: 'alice',
        tasks: ['Review code', 'Write tests', 'Deploy app'],
        sender: 'Bot Assistant'
      };

      const result = await processor.processTemplate(template, variables);

      expect(result).toContain('Hello, Alice!');
      expect(result).toContain('Here are your tasks (3 total):');
      expect(result).toContain('  1. Review code');
      expect(result).toContain('  2. Write tests');
      expect(result).toContain('  3. Deploy app');
      expect(result).toContain('Best regards,');
      expect(result).toContain('Bot Assistant');
    });
  });
});