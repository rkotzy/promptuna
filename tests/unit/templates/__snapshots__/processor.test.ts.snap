// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TemplateProcessor > complex template scenarios > should handle complex message formatting 1`] = `
"Hello, Alice!


Here are your tasks (3 total):
  1. Review code  2. Write tests  3. Deploy app


Best regards,
Bot Assistant"
`;
